### PR TITLE
Create ProfileCreator Service Object

### PR DESCRIPTION
**Pull request by envy99**
**Add ProfileCreator Service Object and ApplicatonService as a parent of all Service Objects**

- ApplicatonService is used for easier access to Service Objects. 
Instead of 
`ServiceObject.new(params).call`
We can use
`ServiceObject.call(params)`

### CODE REVIEW

**Review by kubi-dev — CHANGES_REQUESTED** [review]
You have introduced ServiceObject here but you don't use all of its potential:

1) You still use model callback here. We want to avoid that. I would create a SO to create a user, which includes a profile creation as well and call it from the controller

2) The name convention is `BusinessLogicObject::Action` so I would name it `User::Create`

3) Specs are missing

Great job with `ApplicationService`

To sum up we want to have a service object that creates a user and its profile. And the service object should be tested and called from the controller (instead of user.save)

**Review by jonaszkadziela — COMMENTED** [review]
Remember to always run `rubocop` before pushing changes to GitHub. Your code doesn't adhere to Ragnarson's styling standards.

File: app/services/user/create.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def call
+    create_profile(@user) if save_user(@user)
+    @user
+  end
+
+  protected
+
+  def save_user(user)
```
Comment (kubi-dev) [inline app/services/user/create.rb]: I would not create this method. `@user.save`is quite short (shorter then your method name) and easy to read and follow

File: app/services/user/create.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  def initialize(params)
+    @user = params[:user]
+  end
+
+  def call
+    create_profile(@user) if save_user(@user)
+    @user
+  end
+
+  protected
```
Comment (kubi-dev) [inline app/services/user/create.rb]: What is a difference between `protected` and `private`? Why did you choose `protected` here?

File: app/controllers/users/registrations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+class Users::RegistrationsController < Devise::RegistrationsController
+  before_action :configure_sign_up_params, only: [:create]
+
+  def create
+    resource = User::Create.call(user: build_resource(sign_up_params))
```
Comment (kubi-dev) [inline app/controllers/users/registrations_controller.rb]: I would pass just params as an argument. The controller knows about params and can have no idea what should be built from them. It a service object role to know what to build and how to do this.

👍 for use of key arguments

File: app/services/user/create.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+  def save_user(user)
+    user.save
+  end
+
+  def create_profile(user)
```
Comment (kubi-dev) [inline app/services/user/create.rb]: You don't need to pass an argument to methods as you have `@user` instance variable which is accessible from any instance method. 

To make it nicer we can create a getter `attr_reader :user`

File: spec/features/profile/profile_create_spec.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require "rails_helper"
+
+RSpec.feature "Creating a profile" do
+  describe "after user registration" do
+    scenario "deletes his profile" do
```
Comment (kubi-dev) [inline spec/features/profile/profile_create_spec.rb]: I think the scenario description is not valid. We are creating his profile, aren't we?

**Review by kubi-dev — CHANGES_REQUESTED** [review]
Nice work. I wrote some comments with small suggestions for improvements.

Nice feature spec! Can we also write a SO spec? I know it is covered by feature case but we can learn more writing also spec for the service object. Maybe we can cover any edge cases there?

File: app/services/user/create.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+class User::Create < ApplicationService
+  attr_accessor :user
+
+  def initialize(params)
+    @user = params[:user]
```
Comment (kubi-dev) [inline app/services/user/create.rb]: I would name it `@params` because '@user` is confusing that it is already an object

File: app/controllers/users/registrations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+class Users::RegistrationsController < Devise::RegistrationsController
+  before_action :configure_sign_up_params, only: [:create]
```
Comment (jonaszkadziela) [inline app/controllers/users/registrations_controller.rb]: According to Ragnarson's Ruby style guide, I would change `[:create]` to `%i(create)`.

File: app/controllers/users/registrations_controller.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    if resource.persisted?
+      handle_login(resource)
+    else
+      clean_up_passwords resource
+      set_minimum_password_length
+      respond_with resource
+    end
+  end
+
+  protected
```
Comment (jonaszkadziela) [inline app/controllers/users/registrations_controller.rb]: I suggest changing `protected` to `private`. It does not really matter in that case, but I would do it just to be consistent with the rest of our controllers.

File: app/controllers/users/registrations_controller.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      respond_with resource, location: after_sign_up_path_for(resource)
+    else
+      set_flash_message! :notice, :"signed_up_but_#{resource.inactive_message}"
+      expire_data_after_sign_in!
+      respond_with resource, location: after_inactive_sign_up_path_for(resource)
+    end
+  end
+
+  def configure_sign_up_params
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:attribute])
```
Comment (jonaszkadziela) [inline app/controllers/users/registrations_controller.rb]: Same here - According to Ragnarson's Ruby style guide, I would change [:attribute] to %i(attribute).

File: spec/services/user/create_spec.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+require "rails_helper"
+
+RSpec.describe User::Create do
+  let(:sign_up_params) do
+    { email: "test@example.com", password: "example", password_confirmation: "example" }
```
Comment (jonaszkadziela) [inline spec/services/user/create_spec.rb]: I would do it that way: `let(:sign_up_params) { attributes_for(:user) }`

**Review by jonaszkadziela — COMMENTED** [review]
I pointed out a few things that could have been done better. You may implement those suggestions if you want, but they are not obligatory by any means 😉

File: app/services/user/create.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+class User::Create < ApplicationService
+  attr_accessor :user
+  attr_reader   :params
```
Comment (kubi-dev) [inline app/services/user/create.rb]: I would move both of them to private section

File: app/services/user/create.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
+class User::Create < ApplicationService
+  def initialize(params)
+    @params = params[:user]
+  end
+
+  def call
+    self.user = build_user
```
Comment (kubi-dev) [inline app/services/user/create.rb:7]: We can get rid of this self.

**Review by kubi-dev — APPROVED** [review]

File: spec/features/profile/profile_create_spec.rb [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+RSpec.feature "Creating a profile" do
+  describe "after user registration" do
+    scenario "create his profile" do
+      visit new_user_registration_path
+      fill_in "user_email", with: "test@example.com"
+      fill_in "user_password", with: "password"
+      fill_in "user_password_confirmation", with: "password"
+      find("input[name='commit']").click
+      within("//div[@id='navMenu']") do
```
Comment (jonaszkadziela) [inline spec/features/profile/profile_create_spec.rb]: We have changed the ids naming scheme.
Now instead of camel case, we use underscores 😉 

Change `within("//div[@id='navMenu']")` to `within("#nav_menu")` so as to adhere to the new naming convention.

**Comment by kubi-dev** [comment]
Can't wait to see it merged ;)
